### PR TITLE
feat: Make sure all active sessions are cleaned up properly after a termination signal is received

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
## Proposed changes

We need to make sure all sessions are terminated properly if one terminates Appium process.
Related to https://github.com/appium/appium-xcuitest-driver/pull/1153#issuecomment-582005789

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
@KazuCocoa Would you mind checking https://github.com/appium/appium/pull/13913 instead?

**Review by jlipps — APPROVED** [review]
THANK YOU this has been bothering me for some time

File: lib/appium.js [hunk 1]
```diff
@@ -490,6 +484,30 @@ class AppiumDriver extends BaseDriver {
...
+    log.debug(`Cleaning up ${_.size(this.sessions)} active session` +
+      (_.size(this.sessions) === 1 ? '' : 's'));
+    const cleanupPromises = force
+      ? _.values(this.sessions).map((drv) => drv.startUnexpectedShutdown(reason && new Error(reason)))
+      : _.keys(this.sessions).map((id) => this.deleteSession(id));
+    for (const cleanupPromise of cleanupPromises) {
+      try {
+        await cleanupPromise;
+      } catch (e) {
+        log.debug(e);
```
Comment (jlipps) [inline lib/appium.js]: do we want to exit with code 1 in this case as well? right now any errors during cleanup are ignored

Comment (imurchie) [reply lib/appium.js]: I think if this is going to do that, it ought to note the error and then return whether it errored or not, so all the promises can be awaited.

File: lib/appium.js [hunk 2]
```diff
@@ -490,6 +484,30 @@ class AppiumDriver extends BaseDriver {
...
+    const {
+      force = false,
+      reason = null,
+    } = opts;
+    log.debug(`Cleaning up ${_.size(this.sessions)} active session` +
+      (_.size(this.sessions) === 1 ? '' : 's'));
+    const cleanupPromises = force
+      ? _.values(this.sessions).map((drv) => drv.startUnexpectedShutdown(reason && new Error(reason)))
+      : _.keys(this.sessions).map((id) => this.deleteSession(id));
+    for (const cleanupPromise of cleanupPromises) {
```
Comment (imurchie) [inline lib/appium.js]: Can any of these be done in parallel?

Comment (mykola-mokhnach) [reply lib/appium.js]: they are already done in parallel. Each promise is scheduled as soon as it is added to the array

File: lib/appium.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/appium.js]: we actuallydon't care about the result of each session deletion promise. This whole method is just to make sure there are no pending leftovers

**Comment by KazuCocoa** [comment]
Only this:

```
[Appium] Received SIGTERM - shutting down
[debug] [Appium] Cleaning up 1 active session
[Appium] Closing session, cause was 'The process has received SIGTERM signal'
[Appium] Removing session 'a64ba674-e366-4722-820d-f3dfa13fed41' from our master session list
[DevCon Factory] Releasing connections for 242bfcc998fd156df0bd0ad1dde8ab8e0a032114 device on any port number
[DevCon Factory] Found cached connections to release: ["242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100"]
[DevCon Factory] Releasing the listener for '242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100'
[debug] [DevCon Factory] Cached connections count: 0
[debug] [XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
```

```
kazu$ lsof -i:8100
COMMAND   PID USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
node    92731 kazu   25u  IPv6 0xdf06910ae448d5ed      0t0  TCP *:xprint-server (LISTEN)
node    92731 kazu   29u  IPv4 0xdf06910afb9f70ad      0t0  TCP localhost:62889->localhost:xprint-server (ESTABLISHED)
node    92731 kazu   30u  IPv6 0xdf06910b133ed92d      0t0  TCP localhost:xprint-server->localhost:62889 (ESTABLISHED)
kazu$ kill -SIGINT -92731
kazu$ lsof -i:8100
COMMAND   PID USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
node    92731 kazu   29u  IPv4 0xdf06910afb9f70ad      0t0  TCP localhost:62889->localhost:xprint-server (ESTABLISHED)
node    92731 kazu   30u  IPv6 0xdf06910b133ed92d      0t0  TCP localhost:xprint-server->localhost:62889 (ESTABLISHED)
kazu$ 
```


This +  appium/appium-xcuitest-driver#1153
```
[Appium] Received SIGINT - shutting down
[debug] [Appium] Cleaning up 1 active session
[Appium] Closing session, cause was 'The process has received SIGINT signal'
[Appium] Removing session 'cd33d6e2-7add-4cd3-b360-11cbeee3f58c' from our master session list
[DevCon Factory] Releasing connections for 242bfcc998fd156df0bd0ad1dde8ab8e0a032114 device on any port number
[DevCon Factory] Found cached connections to release: ["242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100"]
[DevCon Factory] Releasing the listener for '242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100'
[debug] [DevCon Factory] Cached connections count: 0
[debug] [XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
[iProxy@242bfcc9] The connection has been closed
```

```
kazu$ lsof -i:8100
COMMAND   PID USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
node    93154 kazu   25u  IPv6 0xdf06910b133ed92d      0t0  TCP *:xprint-server (LISTEN)
node    93154 kazu   29u  IPv4 0xdf06910b195130ad      0t0  TCP localhost:62918->localhost:xprint-server (ESTABLISHED)
node    93154 kazu   30u  IPv6 0xdf06910b133ef02d      0t0  TCP localhost:xprint-server->localhost:62918 (ESTABLISHED)
kazu$ 
kazu$ 
kazu$ kill -SIGINT -93154
kazu$ 
kazu$ lsof -i:8100
```

Seems good to me 🎉

**Comment by KazuCocoa** [comment]
Below is another result after creating multiple sessions (without quitting them)

```
[HTTP]
[Appium] Received SIGINT - shutting down
[debug] [Appium] Cleaning up 3 active sessions
[Appium] Closing session, cause was 'The process has received SIGINT signal'
[Appium] Removing session 'cbf0469f-5c12-4d33-b5cb-95a1e853359c' from our master session list
[Appium] Closing session, cause was 'The process has received SIGINT signal'
[Appium] Removing session 'ec04355b-3947-44f1-93b3-761c5f1836fa' from our master session list
[Appium] Closing session, cause was 'The process has received SIGINT signal'
[Appium] Removing session 'fae64007-3514-4242-a2a5-cab7615ff935' from our master session list
[DevCon Factory] Releasing connections for 242bfcc998fd156df0bd0ad1dde8ab8e0a032114 device on any port number
[DevCon Factory] Found cached connections to release: ["242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100"]
[DevCon Factory] Releasing the listener for '242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100'
[DevCon Factory] Releasing connections for 242bfcc998fd156df0bd0ad1dde8ab8e0a032114 device on any port number
[DevCon Factory] Found cached connections to release: ["242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100"]
[DevCon Factory] Releasing the listener for '242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100'
[DevCon Factory] Releasing connections for 242bfcc998fd156df0bd0ad1dde8ab8e0a032114 device on any port number
[DevCon Factory] Found cached connections to release: ["242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100"]
[DevCon Factory] Releasing the listener for '242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100'
[debug] [DevCon Factory] Cached connections count: 0
[debug] [DevCon Factory] Cached connections count: 0
[debug] [DevCon Factory] Cached connections count: 0
[debug] [XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
[debug] [XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
[debug] [XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
[iProxy@242bfcc9] The connection has been closed
[iProxy@242bfcc9] The connection has been closed
[iProxy@242bfcc9] The connection has been closed
```

**Comment by KazuCocoa** [comment]
Android and iOS mixed

```
[Appium] Received SIGINT - shutting down
[debug] [Appium] Cleaning up 2 active sessions
[Appium] Closing session, cause was 'The process has received SIGINT signal'
[Appium] Removing session 'e3e4de8f-7a74-4b48-8e77-f66cc19e5aa4' from our master session list
[Appium] Closing session, cause was 'The process has received SIGINT signal'
[Appium] Removing session '41b8700b-cd78-4843-929b-25586fa548c5' from our master session list
[debug] [UiAutomator2] Deleting UiAutomator2 session
[DevCon Factory] Releasing connections for 242bfcc998fd156df0bd0ad1dde8ab8e0a032114 device on any port number
[DevCon Factory] Found cached connections to release: ["242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100"]
[DevCon Factory] Releasing the listener for '242bfcc998fd156df0bd0ad1dde8ab8e0a032114:8100'
[debug] [UiAutomator2] Deleting UiAutomator2 server session
[debug] [WD Proxy] Matched '/' to command name 'deleteSession'
[debug] [WD Proxy] Proxying [DELETE /] to [DELETE http://127.0.0.1:8200/wd/hub/session/91acec0c-3f8f-42b3-af3c-8f49c0b7572d] with no body
[iProxy@242bfcc9] The connection has been closed
[debug] [DevCon Factory] Cached connections count: 0
[debug] [XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
[Logcat] Logcat terminated with code null, signal SIGINT
[debug] [Instrumentation] The process has exited with code null
[debug] [WD Proxy] Got response with status 200: {"sessionId":"91acec0c-3f8f-42b3-af3c-8f49c0b7572d","value":null}
[debug] [UiAutomator2] Resetting IME to 'io.appium.settings/.UnicodeIME'
[debug] [ADB] Running '/Users/kazu/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell ime set io.appium.settings/.UnicodeIME'
[debug] [ADB] Running '/Users/kazu/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell am force-stop io.appium.android.apis'
[debug] [Logcat] Stopping logcat capture
[debug] [Logcat] Logcat already stopped
[debug] [ADB] Removing forwarded port socket connection: 8200
[debug] [ADB] Running '/Users/kazu/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 forward --remove tcp\:8200'
[UiAutomator2] Restoring hidden api policy to the device default configuration
[debug] [ADB] Running '/Users/kazu/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell settings delete global hidden_api_policy_pre_p_apps'
[debug] [ADB] Running '/Users/kazu/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell settings delete global hidden_api_policy_p_apps'
[debug] [ADB] Running '/Users/kazu/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell settings delete global hidden_api_policy'
```

**Review by KazuCocoa — APPROVED** [review]
the behaviour looks good to me (with the xcuitest change, too)
Brilliant 🎉

**Comment by mykola-mokhnach** [comment]
Many thanks for your help @KazuCocoa

**Pull request merged**